Source/Core/DolphinQt2/Settings/USBDeviceAddToWhitelistDialog.cpp

```diff
@@ -0,0 +1,169 @@
+          &USBDeviceAddToWhitelistDialog::OnRefreshDevicesTimer);
+  m_refresh_devices_timer->start(1000);
+
+  main_layout->addWidget(usb_inserted_devices_list);
+  main_layout->addWidget(whitelist_buttonbox);
+
+  // i18n: VID means Vendor ID (in the context of a USB device)
+  device_vid_textbox->setPlaceholderText(tr("Device VID (e.g., 057e)"));
+  // i18n: PID means Product ID (in the context of a USB device), not Process ID
+  device_pid_textbox->setPlaceholderText(tr("Device PID (e.g., 0305)"));
+}
+
+void USBDeviceAddToWhitelistDialog::RefreshDeviceList()
+{
+  const auto& current_devices = USBUtils::GetInsertedDevices();
+  if (current_devices == m_shown_devices)
+    return;
+  QSignalBlocker blocker(usb_inserted_devices_list);  // unblocks when out of scope.
+  const auto selection_string = usb_inserted_devices_list->currentItem();
+  usb_inserted_devices_list->clear();
+  for (const auto& device : current_devices)
+  {
+    if (SConfig::GetInstance().IsUSBDeviceWhitelisted(device.first))
+      continue;
+    usb_inserted_devices_list->addItem(QString::fromStdString(device.second));
+  }
+
+  usb_inserted_devices_list->setCurrentItem(selection_string);
+
+  m_shown_devices = current_devices;
+}
+
+void USBDeviceAddToWhitelistDialog::AddUSBDeviceToWhitelist()
+{
+  const std::string vid_string = StripSpaces(device_vid_textbox->text().toStdString());
+  const std::string pid_string = StripSpaces(device_pid_textbox->text().toStdString());
+  if (!IsValidUSBIDString(vid_string))
+  {
+    // i18n: Here, VID means Vendor ID (for a USB device).
+    QErrorMessage* error = new QErrorMessage();
+    error->setWindowTitle(tr("USB Whitelist Error"));
+    error->showMessage(tr("The entered VID is invalid."));
+    return;
+  }
+  if (!IsValidUSBIDString(pid_string))
+  {
+    // i18n: Here, PID means Product ID (for a USB device).
+    QErrorMessage* error = new QErrorMessage();
+    error->setWindowTitle(tr("USB Whitelist Error"));
+    error->showMessage(tr("The entered PID is invalid."));
+    return;
+  }
+
+  const u16 vid = static_cast<u16>(std::stoul(vid_string, nullptr, 16));
+  const u16 pid = static_cast<u16>(std::stoul(pid_string, nullptr, 16));
+
+  if (SConfig::GetInstance().IsUSBDeviceWhitelisted({vid, pid}))
+  {
+    QErrorMessage* error = new QErrorMessage();
+    error->showMessage(tr("This USB device is already whitelisted."));
+    return;
+  }
+  SConfig::GetInstance().m_usb_passthrough_devices.emplace(vid, pid);
+  SConfig::GetInstance().SaveSettings();
+  // Not the nicest way of doing this but...
+  WiiPane* p = static_cast<WiiPane*>(parent());
+  p->PopulateUSBPassthroughListWidget();
+  close();
+}
+
+void USBDeviceAddToWhitelistDialog::OnRefreshDevicesTimer()
+{
+  RefreshDeviceList();
+}
+
+void USBDeviceAddToWhitelistDialog::OnDeviceSelection()
+{
+  // Not the nicest way of doing this but...
+  QString device = usb_inserted_devices_list->currentItem()->text().left(9);
```

Comment:
See my comment on the device list in WiiPane on how to do this in a non-fragile way using the data field of QListWidgetItems.